Ilariae/wormhole settlements (#254)
wormhole-foundation/wormhole-docs: build/contract-integrations/settlements/solver.md

---
title: Wormhole Settlements
description: TODO
---

# Running a Wormhole Settlement Solver

This page provides instructions on how to set up, configure, and run a Solver for Wormhole Settlement using the [example solver](https://github.com/wormholelabs-xyz/example-liquidity-layer/tree/update-solver-example/solver){target=\_blank}.

## Understanding the Hub: Solana Matching Engine

A central component of the protocol architecture is the permissionless English auction conducted on Solana, specifically the Matching Engine contract, allowing any third-party solver to interact with the matching engine to place bids or improve existing ones. The Matching Engine Contract includes four key instructions:

- `initialize_auction` - creates a new auction account on-chain. This sets basic parameters like the auction's token mint, the amount required, and the bidding period details

> - `bid` - allows a solver to place or update a bid on the active auction. The solver transfers the bid amount to the program escrow account, which ensures they actually have the liquidity. With each successful call of `bid`, the program updates the auction to the new highest bidder and the prior bid is atomically sent back to the originating solver, allowing that solver to repurpose those funds if it so chooses to improve the bid once more

Reviewer: ```suggestion
- `bid` - allows a solver to place or update a bid on the active auction. The solver transfers the bid amount to the program escrow account, which ensures they have liquidity. With each successful call of `bid`, the program updates the auction to the new highest bidder, and the prior bid is atomically sent back to the originating solver, allowing that solver to repurpose those funds if it so chooses to improve the bid once more
```

- Facilitating the actual cross-chain transfer by locking/burning assets on Solana and minting/unlocking them on the destination chain
- Rebalancing once the origin chain transaction finalizes and is redeemed back on Solana

For information on how the protocol functions and its core features, please visit the [Wormhole Settlement](/docs/learn/messaging/wormhole-settlements/overview/){target=\_blank} page.

### Testnet Example Solver

The following is an example solver to fulfill fast orders by interacting with the Matching Engine on Solana.

!!!important
    This example is not optimized for performance and has only been tested on Solana devnet. Any assumptions made in this example may not translate to mainnet.

#### Getting Started

In order to build and install dependencies locally in this repo, you will need `node v20.18.1` and `npmv. Get started by installing `nvm` using this [installation guide](https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating){target=\_blank}.

Run the command below to set up your environment. This installs node dependencies and the Matching Engine package.

```sh
make dependencies
```

#### Setting up Config

> Here is an example `config.json` file for Solana devnet. The keys here are required for both the publisher and Example Solver processes.

Reviewer: ```suggestion
Here is an example of a `config.json` file for Solana devnet. The keys here are required for both the publisher and Example Solver processes.
```

2025-01-21 16:40:24.890 [publisher] info: Fast VAA. chain=OptimismSepolia, sequence=33639, vaaTime=1737499224
```

#### Running the Example Solver

Using the same config for your publisher, run the example solver with the command below.

```sh
CONFIG=path/to/config.json make run-solver
```

**We recommend writing log output to a file so errors can be tracked.** The example config above specifies an example log filename.
This process reads the following environment variables:

```sh
SOLANA_PRIVATE_KEY_1=
SOLANA_PRIVATE_KEY_2=
SOLANA_PRIVATE_KEY_3=
SOLANA_PRIVATE_KEY_4=
SOLANA_PRIVATE_KEY_5=
```

At least one of these environment variables must be defined as a keypair encoded in base64 format. These payers must have SOL to send transactions on Solana devnet. If they need funds, they can request them from the [Solana devnet faucet](https://faucet.solana.com/){target=\_blank}.

The example solver assumes that these payers own USDC Associated Token Accounts, which will be used to fulfill fast transfers. These ATAs must be funded with Solana Devnet USDC. If your ATAs need funds, request some at the [Circle testnet faucet](https://faucet.circle.com/){target=\_blank}.

> Wallets and their corresponding ATA will be disabled if there are not enough funds to pay for transactions or fulfill fast transfers. These constraints can be modified using the `updatePayerMinimumLamports` and `updateTokenMinimumBalance` methods.

Reviewer: ```suggestion
Wallets and their corresponding ATA will be disabled if there are insufficient funds to pay for transactions or fulfill fast transfers. These constraints can be modified using the `updatePayerMinimumLamports` and `updateTokenMinimumBalance` methods.
```

An address lookup table is required to execute some transactions. Use the command below to create one.

```sh
CONFIG=path/to/config.json make create-lut
```

> `SOLANA_PRIVATE_KEY_1` must be defined in order for this script to work.

Reviewer: ```suggestion
`SOLANA_PRIVATE_KEY_1` must be defined for this script to work.
```

For information on how the protocol functions and its core features, please visit the [Wormhole Settlement](/docs/learn/messaging/wormhole-settlements/overview/){target=\_blank} page.

## Background

The Solana Matching Engine's permissionless English auction is a central component of Wormhole Settlements protocol architecture. The Matching Engine contract allows any third-party solver to interact with the matching engine to place bids or improve existing ones. The contract includes four key instructions:

1. `initialize_auction` - creates a new auction account on-chain and sets basic parameters like the auction's token mint, the amount required, and the bidding period details
2. `bid` - allows a solver to place or update a bid on the active auction

> 3. `finalize_auction` - following the conclusion of the auction, this instruction completes the fast transfer by sending funds to the recipient on the target chain. This instruction may call the Circle CCTP contract or release an NTT contract in the future, depending on the shuttle asset in question. Failure to execute this message within a predefined grace period may result in a penalty for the winning bidder.

Reviewer: ```suggestion
3. `finalize_auction` - following the conclusion of the auction, this instruction completes the fast transfer by sending funds to the recipient on the target chain. Depending on the shuttle asset in question, this instruction may call the Circle CCTP contract or release an NTT contract in the future
```

If the winning solver fails to call the `finalize_auction` instruction, other competing solvers may permissionlessly 'slash' the solver by executing the instruction on their behalf and collecting a portion of the original security deposit as a reward. The remaining portion is routed to the user as compensation for the unanticipated delay. This mechanism properly incentivizes timely execution through solver redundancy and competition.

## Testnet Example Solver

You can clone the Wormhole [`example-liquidity-layer`](https://github.com/wormholelabs-xyz/example-liquidity-layer){target=\_blank} repository to use the included [`solver`](https://github.com/wormholelabs-xyz/example-liquidity-layer/tree/main/solver){target=\_blank} directory as an example solver to fulfill fast orders by interacting with the Matching Engine on Solana.

!!!warning

>     This example is not optimized for performance, has only been tested on Solana devnet, and is not intended for production use. Any assumptions made in this example may not translate to mainnet.

Reviewer: ```suggestion
    This example is not optimized for performance. It has only been tested on Solana devnet and is not intended for production use. Any assumptions made in this example may not translate into mainnet.
```

2025-01-21 16:40:24.890 [publisher] info: Fast VAA. chain=OptimismSepolia, sequence=33639, vaaTime=1737499224
```

To set up the Pyth Beacon (which is run using make `run-publisher`), you may need to increase the UDP buffer size for the OS:

```sh
# for linux
sudo sysctl -w net.core.rmem_max=2097152
sudo sysctl -w net.core.rmem_default=2097152
# for macos
sudo sysctl -w net.inet.udp.recvspace=2097152
```

### Running the Example Solver

Using the same config for your publisher, run the example solver with the command below.

```sh
CONFIG=path/to/config.json make run-solver
```

> It is recommended you write log output to a file so errors can be tracked. The example config above specifies an example log filename.

Reviewer: ```suggestion
It is recommended that you write log output to a file so errors can be tracked. The example config above specifies an example log filename.
```

> At least one of these environment variables must be defined as a keypair encoded in base64 format. These payers must have SOL to send transactions on Solana devnet. If they need funds, they can request them from the [Solana devnet faucet](https://faucet.solana.com/){target=\_blank}.

Reviewer: ```suggestion
At least one of these environment variables must be defined as a keypair encoded in base64 format. These payers must have SOL to send transactions on Solana devnet. They can request funds from the [Solana devnet faucet](https://faucet.solana.com/){target=\_blank} if they need them.
```
